debugger/unix_tools.ml

```diff
@@ -22,24 +22,27 @@ open Unix
 
 (*** Convert a socket name into a socket address. ***)
 let convert_address address =
-  try
-    let n = String.index address ':' in
-      let host = String.sub address 0 n
-      and port = String.sub address (n + 1) (String.length address - n - 1)
-      in
-        (PF_INET,
-         ADDR_INET
-           ((try inet_addr_of_string host with Failure _ ->
-               try (gethostbyname host).h_addr_list.(0) with Not_found ->
-                 prerr_endline ("Unknown host: " ^ host);
-                 failwith "Can't convert address"),
-            (try int_of_string port with Failure _ ->
-               prerr_endline "The port number should be an integer";
-               failwith "Can't convert address")))
-  with Not_found ->
-    match Sys.os_type with
-      "Win32" -> failwith "Unix sockets not supported"
-    | _ -> (PF_UNIX, ADDR_UNIX address)
+  if String.length address == 0 then
+    failwith "Can't convert address: empty or invalid address";
+  let is_likely_ipv6 str n =
+    n >= 4 && str.[0] = '[' && str.[n] = ']' in
+  let un_addr =
+    { ai_family = PF_UNIX; ai_socktype = SOCK_STREAM; ai_protocol = 0;
+      ai_addr = ADDR_UNIX address; ai_canonname = ""; } in
+  match String.rindex address ':' with
+  | exception Not_found -> un_addr
+  | n ->
+     let likely_ipv6 = is_likely_ipv6 address (n - 1) in
+     if not (Filename.is_implicit address) && not likely_ipv6 then
```

Comment:
I find the logic in this block of code too hard to follow. Before your change the idea was rather clear: either there is a `:` and we handle this as an IP (V4) address, or there isn't and we assume a unix socket path. Now what is the logic, and is it apparent in the code?

For example I think that it is not clear what `not (Filename.is_implicit ...)` is testing for, and maybe this could be helped with variable naming: `let is_an_explicit_path = not (Filename.is_implicit address) in ...`.

Reply:
Well a path like `C:\TMP\my.sock` or `./my:sock` is obviously not a `host:port`. `not (Filename.is_implicit ...)` caches paths that are rooted or relative to the current directory.
If we only look at whether the address contains a `:` to decide whether it's an Unix domain socket or an inet address, then we'll miss absolute paths on Windows.
I'll look for a simpler compromise.

Reply:
I'm not saying that the previous logic was better -- I agree it was worse, especially on Windows. I am asking for a clear description of what the new logic is, and code that matches this clear description.